drivers/peci/peci_handlers.c

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2020 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <drivers/peci.h>
+#include <syscall_handler.h>
+
+static inline int z_vrfy_peci_config(struct device *dev, u32_t bitrate)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_PECI));
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, config));
+
+	return z_impl_peci_config((struct device *)dev, (u16_t)bitrate);
+}
+#include <syscalls/peci_config_mrsh.c>
+
+static inline int z_vrfy_peci_enable(struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_PECI));
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, enable));
+
+	return z_impl_peci_enable((struct device *)dev);
+}
+#include <syscalls/peci_enable_mrsh.c>
+
+static inline int z_vrfy_peci_disable(struct device *dev)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_PECI));
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, disable));
+
+	return z_impl_peci_disable((struct device *)dev);
+}
+#include <syscalls/peci_disable_mrsh.c>
+
+static inline int z_vrfy_peci_transfer(struct device *dev,
+				       const struct peci_msg *msg)
+{
+	Z_OOPS(Z_SYSCALL_OBJ(dev, K_OBJ_DRIVER_PECI));
+	Z_OOPS(Z_SYSCALL_DRIVER_PECI(dev, transfer));
+
+	Z_OOPS(Z_SYSCALL_MEMORY_READ((const struct peci_msg *)msg,
```

Comment:
need to make a copy and validate members of that. otherwise vulnerable to TOCTOU attacks.
use z_user_from_copy() here and copy everything to an instance on the stack. and then the following calls should work on the copy, which will be passed to the implementation.

Reply:
@andrewboie  update the stack copy, point me to some driver example for the additional validation of the members

Reply:
you need something like:

Z_OOPS(z_user_from_copy(&msg_copy, msg, sizeof(msg_copy)));
Z_OOPS(Z_SYSCALL_MEMORY_WRITE(msg_copy.rx_buffer->buf, msg_copy.rx_buffer->len));
Z_OOPS(Z_SYSCALL_MEMORY_READ(msg_copy.tx_buffer->buf, msg_copy.tx_buffer->len));

For any struct passed in from user mode you have to consider it potentially hostile and its contents changed at any time. So you must make a copy of it and use the copy.

Reply:
@andrew got it, if that's the case it seems I already covered in last update.

Reply:
no, you haven't unfortunately.
for one thing, you are doing Z_SYSCALL_MEMORY_READ() on the original msg and not the copy
you need to do the z_user_from_copy() before looking at ANY of its members.